src/Common/tests/Cryptography/CryptoUtils.cs

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Security.Cryptography;
+
+namespace Test.Cryptography
+{
+    internal static class CryptoUtils
+    {
+        internal static byte[] Encrypt(this SymmetricAlgorithm alg, byte[] plainText, int blockSizeMultipler = 1)
+        {
+            return alg.CreateEncryptor().Transform(plainText, blockSizeMultipler);
+        }
+
+        internal static byte[] Decrypt(this SymmetricAlgorithm alg, byte[] cipher, int blockSizeMultipler = 1)
+        {
+            return alg.CreateDecryptor().Transform(cipher, blockSizeMultipler);
+        }
+
+        internal static byte[] Transform(this ICryptoTransform transform, byte[] input, int blockSizeMultipler = 1)
+        {
+            List<byte> output = new List<byte>(input.Length);
+            int blockSize = transform.InputBlockSize * blockSizeMultipler;
+            for (int i = 0; i <= input.Length; i += blockSize)
+            {
+                int count = Math.Min(blockSize, input.Length - i);
+                if (count >= blockSize)
+                {
+                    byte[] buffer = new byte[blockSize];
+                    int numBytesWritten = transform.TransformBlock(input, i, count, buffer, 0);
+                    Array.Resize(ref buffer, numBytesWritten);
+                    output.AddRange(buffer);
+                }
+                else
+                {
+                    byte[] finalBlock = transform.TransformFinalBlock(input, i, count);
```

Comment:
“I” is the loop induction variable, not “blockSize”.

The multiplier controls the size of the chunks we feed to Transform(). Purpose is to let us test multiple block transforms, and edge cases (like never calling Transform but only TransformFinalBlock.)


From: Jeremy Barton [mailto:notifications@github.com]
Sent: Monday, October 5, 2015 9:54 AM
To: dotnet/corefx <corefx@noreply.github.com>
Cc: Atsushi Kanamori <atsushik@microsoft.com>
Subject: Re: [corefx] Implement AesCng and TripleDESCng (#3640)


In src/Common/tests/Cryptography/CryptoUtils.cs<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcorefx%2fpull%2f3640%23discussion_r41167902&data=01%7c01%7catsushik%40microsoft.com%7c54180f2c475742db82e408d2cda593a9%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=848NLb6LHMulml5yt8WSIzY9mBp40XJmHB%2fb67rTnLE%3d>:

> +        {

> +            List<byte> output = new List<byte>(input.Length);

> +            int blockSize = transform.InputBlockSize * blockSizeMultipler;

> +            for (int i = 0; i <= input.Length; i += blockSize)

> +            {

> +                int count = Math.Min(blockSize, input.Length - i);

> +                if (count >= blockSize)

> +                {

> +                    byte[] buffer = new byte[blockSize];

> +                    int numBytesWritten = transform.TransformBlock(input, i, count, buffer, 0);

> +                    Array.Resize(ref buffer, numBytesWritten);

> +                    output.AddRange(buffer);

> +                }

> +                else

> +                {

> +                    byte[] finalBlock = transform.TransformFinalBlock(input, i, count);

I guess I don't understand what blockSizeMultiplier is supposed to be doing.

Transform("hi".ToBytes(), 10) is the same as Transform("hi".ToBytes(), 1), in that both just call TransformFinalBlock.

And since blocksize isn't decremented during looping, Transform(new byte[4096], 1) is just going to call TransformBlock over and over, but never TransformFinalBlock.

—
Reply to this email directly or view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fdotnet%2fcorefx%2fpull%2f3640%2ffiles%23r41167902&data=01%7c01%7catsushik%40microsoft.com%7c54180f2c475742db82e408d2cda593a9%7c72f988bf86f141af91ab2d7cd011db47%7c1&sdata=1yeyCHPNANMeU2GkeSAobIMCJqd1t8cDPJoi5omIf%2b8%3d>.